Monster looting improvements

NullCGT opened the pull request:
- Implemented cursed bag of holding behavior for when monsters loot a cursed bag of holding.
- Allow monsters to loot containers at their location instead of picking them up.

I avoided adding behavior for monsters unlocking containers. It would not be difficult to add, but the number of times we would need to iterate through a monster's inventory for such a niche behavior was difficult for me to justify.

Commit:
Handle monsters looting cursed bags of holding.

Resolves a FIXME in muse.c. Does not yield any messages, since the
player should not be able to see when something happens inside
someone else's bag of holding. Adding messages is possible, but
would probably require adding a completely new function specifically
for monsters looting cursed bags of holding, which feels like
overkill to me.

Commit:
Fix typo in muse.c.

Commit:
Monsters search through containers at their location.

Monsters may still end up wasting turns picking up containers, but
now it should at least be possible for them to loot a container
at their location without picking it up.

Commit:
Improve check on containers at monster location.

entrez commented on src/muse.c:
```diff
@@ -2055,6 +2060,12 @@ mloot_container(
+    /* Monster rummaged through a bag but found nothing of interest, possibly because
+       some items vanished. */
+    if (!res && vismon) {
+        pline("%s %s inside %s.", Monnam(mon), mon->mcansee ? "looks" : "feels about", 
```
Formatting thing, this line and the above comment is longer than 80 characters

NullCGT replied:
Ugh, thanks. I'm not in my usual editor.

entrez commented on src/pickup.c:
```diff
@@ -2608,7 +2607,7 @@ mbag_item_gone(int held, struct obj *item, boolean silent)
     }
 
-    if (*u.ushops && (shkp = shop_keeper(*u.ushops)) != 0) {
+    if (*u.ushops && (shkp = shop_keeper(*u.ushops)) != 0 && user == &g.youmonst) {
```
Also longer than 80 characters

entrez commented on src/muse.c:
```diff
@@ -1957,9 +1967,8 @@ mloot_container(
-    /* FIXME: handle cursed bag of holding */
     if (Is_mbag(container) && container->cursed)
-        return res; /* 0 */
+        boh_loss(mon, container, container->ocarry == mon);
```
Is there anything to stop a monster from repeatedly looking into a bag of holding without "learning" it's cursed and losing all the items?

NullCGT replied:
No. It's a similar problem to monsters not being able to remember whether something is locked.

copperwater replied:
Since monsters implicitly know items as if they were fully identified, shouldn't they respond to a cursed bag of holding by dumping out all the contents so as not to lose items looking in each time?

entrez commented on src/muse.c:
```diff
@@ -1829,6 +1831,17 @@ find_misc(struct monst* mtmp)
+                && Has_contents(otmp) && !otmp->olocked && !otmp->otrapped) {
+                g.m.has_misc = MUSE_BAG;
+                g.m.misc = otmp;
+                return TRUE;
```
Why does a container being on the monster's current spot override any of the other misc items it can use half of the time?  It feels to me like maybe the `return TRUE` shouldn't be here, so that the other rules for things that the monster actually has a need to use can take precedence further down, and the container looting will only happen as a "well, nothing else interesting to do" thing.  That's basically how bags in inventory work, right?  They are the lowest precedence action in `find_misc`.  Is the idea that it should take precedence because the monster might move off the spot and lose the chance to open it?

entrez commented on src/muse.c:
```diff
@@ -1829,6 +1831,17 @@ find_misc(struct monst* mtmp)
+                return TRUE;
+            }
+        }
+    }
```
Does placing this above the `if (nohands(mdat)) return 0;` mean that monsters without hands can still open a container on their location?

Commit:
Further monster looting refinement.

Prevent monsters without hands from looting items at their location, and
stop monsters from prioritizing looting a container at their location
above all else.

entrez commented on src/muse.c:
```diff
@@ -1983,6 +1991,8 @@ mloot_container(
         Strcpy(mpronounbuf, mhe(mon));
     }
+    Strcpy(contnr_nam, an(nearby ? xname(container)
+                                         : distant_name(container, xname)));
```
The '?' and ':' should probably line up like they used to

entrez commented on src/muse.c:
```diff
@@ -2055,6 +2059,13 @@ mloot_container(
+    /* Monster rummaged through a bag but found nothing of interest,possibly
+       because some items vanished. */
+    if (!res && vismon) {
+        pline("%s %s inside %s.", Monnam(mon), 
+                                  mon->mcansee ? "looks" : "feels about",
+                                  contnr_nam);
```
The broken up parts should be aligned with the opening paren of pline, like
```c
        pline("%s %s inside %s.", Monnam(mon), 
              mon->mcansee ? "looks" : "feels about", contnr_nam);
```
or
```c
        pline("%s %s inside %s.", Monnam(mon), 
              mon->mcansee ? "looks" : "feels about",
              contnr_nam);
```
etc

entrez commented on src/pickup.c:
```diff
@@ -2608,7 +2607,8 @@ mbag_item_gone(int held, struct obj *item, boolean silent)
     }
 
-    if (*u.ushops && (shkp = shop_keeper(*u.ushops)) != 0) {
+    if (*u.ushops && (shkp = shop_keeper(*u.ushops)) != 0 
+                      && user == &g.youmonst) {
```
Broken up part should be aligned with the opening paren, like
```c
    if (*u.ushops && (shkp = shop_keeper(*u.ushops)) != 0 
        && user == &g.youmonst) {
```

entrez reviewed (commented):
A couple more formatting nitpicks I wanted to point out.

pat-rankin commented:
This has significant ramifications for bones levels.  If the dead hero was carrying a bag of holding it will be cursed and on the floor.  Depending on where the new hero arrives on the level, any item-using monster in the bag's vicinity could have lots of time to repeatedly loot it before the new character eventually makes it to that location, potentially destroying a lot of items.  (That isn't necessarily a bad thing but most players probably won't like it.)

Right now bags and boxes with lots of stuff are protected by virtue of being too heavy to lift.

copperwater commented:
Wrote this in response to a review comment before seeing it brought up in the main thread:

Since monsters implicitly know items as if they were fully identified, shouldn't they respond to a cursed bag of holding by dumping out all the contents (i.e. tipping it) so as not to lose items looking in each time?

Without something like that, yeah, I can't really get behind allowing monsters to loot bags on the ground.
